### PR TITLE
Prayer bar

### PR DESCRIPTION
**Pull request by Chdata**
Adds a toggle-able prayer bar underneath the health bar, indicating your amount of prayer left.

it can be toggled to show only while praying, only show if your health bar is also visible, or simply be "always on".

image

image

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		return false;
+	}
+
+	@ConfigItem(
+		keyName = "prayerBarFgColor",
+		name = "Prayer Bar FG Color",
+		description = "Color of the overhead prayer bar",
+		position = 3
+	)
+	default Color getPrayerBarColor()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java]: Just use the prayer default color, do not make this configurable, that is just overdoing it

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		return Color.cyan;
+	}
+
+	@ConfigItem(
+		keyName = "prayerBarBgColor",
+		name = "Prayer Bar BG Color",
+		description = "Color of the background of the overhead prayer bar",
+		position = 4
+	)
+	default Color getPrayerBarBackgroundColor()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java]: See my previous comment

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		return Color.white;
+	}
+
+	@ConfigItem(
+		keyName = "prayerBarBgAlpha",
+		name = "Prayer Bar BG Alpha",
+		description = "Translucency of the background of the overhead prayer bar",
+		position = 5
+	)
+	default int getPrayerBarBackgroundAlpha()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java]: We have color pickers. They need fix, this is workaround that should not be present

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		return true;
+	}
+
+	@ConfigItem(
+		keyName = "prayerBarHideIfNonCombat",
+		name = "Hide while out-of-combat",
+		description = "Prayer bar will be hidden while out-of-combat.",
+		position = 2
+	)
+	default boolean hideIfOutOfCombat()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarConfig.java]: Default to true.

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+{
+	private final Client client;
+	private Player localPlayer;
+	private boolean showPrayerBar = true;
+	private final PrayerBarConfig config;
+	private Dimension prayerBarSize = new Dimension(30, 5); // 30x5 is the size of health bars
+	//private static final Logger logger = LoggerFactory.getLogger(PrayerBarOverlay.class);
+
+	@Inject
+	public PrayerBarOverlay(PrayerBarConfig config, Client client)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Private constructor

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+//import org.slf4j.Logger;
+//import org.slf4j.LoggerFactory;
+
+@Singleton
+public class PrayerBarOverlay extends Overlay
+{
+	private final Client client;
+	private Player localPlayer;
+	private boolean showPrayerBar = true;
+	private final PrayerBarConfig config;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Align final variables properly before the non-final ones

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	private Dimension prayerBarSize = new Dimension(30, 5); // 30x5 is the size of health bars
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Mark final

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	//private static final Logger logger = LoggerFactory.getLogger(PrayerBarOverlay.class);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Remove this

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+import net.runelite.api.Client;
+import net.runelite.api.Perspective;
+import net.runelite.api.Player;
+import net.runelite.api.Point;
+import net.runelite.api.Prayer;
+import net.runelite.api.Skill;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.ui.overlay.*;
+//import org.slf4j.Logger;
+//import org.slf4j.LoggerFactory;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Remove comments

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		if (showPrayerBar)
+		{
+			return renderPrayerBar(graphics, localPlayer);
+		}
+		return null;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Add empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	private Dimension renderPrayerBar(Graphics2D graphics, Player localPlayer)
+	{
+		Color backgroundColor = new Color(
+			config.getPrayerBarBackgroundColor().getRed(),
+			config.getPrayerBarBackgroundColor().getGreen(),
+			config.getPrayerBarBackgroundColor().getBlue(),
+			Math.min(config.getPrayerBarBackgroundAlpha(), 255)
+		);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Merge this ); with previous line

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	private boolean isAnyPrayerActive()
+	{
+		for (Prayer pray : Prayer.values()) // Check if any prayers are active
+		{
+			if (client.isPrayerActive(pray))
+			{
+				return true;
+			}
+		}
+		return false;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Add empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarPlugin.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	@Override
+	public Overlay getOverlay()
+	{
+		return overlay;
+	}
+
+	@Subscribe
+	public void onTick(GameTick tick)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarPlugin.java]: The event is named GameTick, so name the method onGameTick

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+	{
+		this.config = config;
+		this.client = client;
+
+		setPosition(OverlayPosition.DYNAMIC);
+		setPriority(OverlayPriority.MED);
+		setLayer(OverlayLayer.UNDER_WIDGETS);
+	}
+
+	public void onTick()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Change this method name to make more sense

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+	//private static final Logger logger = LoggerFactory.getLogger(PrayerBarOverlay.class);
+
+	@Inject
+	public PrayerBarOverlay(PrayerBarConfig config, Client client)
+	{
+		this.config = config;
+		this.client = client;
+
+		setPosition(OverlayPosition.DYNAMIC);
+		setPriority(OverlayPriority.MED);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Why isn't this low

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+		int barWidth = prayerBarSize.width;
+		int barHeight = prayerBarSize.height;
+
+		float ratio = (float) client.getBoostedSkillLevel(Skill.PRAYER) / client.getRealSkillLevel(Skill.PRAYER);
+
+		int progressFill = (int) Math.ceil(Math.min((barWidth * ratio), barWidth)); // Restricted by the width to prevent the bar from being too long while you are boosted above your real prayer level.
+
+		graphics.setColor(Color.white);
+		graphics.fillRect(barX, barY, barWidth, barHeight);
+		graphics.setColor(Color.cyan);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Make this Color.white and Color.cyan constants in this class and name them something like BAR_BG_COLOR and BAR_FG_COLOR

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+import net.runelite.api.Skill;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.ui.overlay.*;
+
+@Singleton
+public class PrayerBarOverlay extends Overlay
+{
+	private final Client client;
+	private final PrayerBarConfig config;
+	private final Dimension prayerBarSize = new Dimension(30, 5); // 30x5 is the size of health bars
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Make this constant, e.g private static final Dimension PRAYER_BAR_SIZE

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+import net.runelite.api.Skill;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.ui.overlay.*;
+
+@Singleton
+public class PrayerBarOverlay extends Overlay
+{
+	private final Client client;
+	private final PrayerBarConfig config;
+	private static final Color BAR_FILL_COLOR = Color.cyan;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Move constants at the top of the class, where they should be

**Comment by Chdata** [comment]
I'll probably still want at least the background to have a configurable color though. I think it's fine as long as the default color is consistent. A dropdown list might be more suited as there were only ~4 colors I tried that looked decent, out of many I tested.

I was going to try using a smaller height for the bar to make more space for other bars to be added and to differentiate this more from the health bar, but it didn't look good on its own that way and would maybe need further non-vanilla stuff like a 1px shadow or something to look nice.

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+	{
+		this.config = config;
+		this.client = client;
+
+		setPosition(OverlayPosition.DYNAMIC);
+		setPriority(OverlayPriority.LOW);
+		setLayer(OverlayLayer.UNDER_WIDGETS);
+	}
+
+	public void checkToShowPrayerBar()
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarOverlay.java]: Generally this type of logic goes into the plugin, and then the overlay checks the plugin state to determine if to draw.

**Comment by Adam-** [comment]
I don't think this is related to prayer flicking at all. Maybe this and #1578 could be merged into a "prayer" plugin?

**Comment by Chdata** [comment]
Yeah, I was thinking it'd make sense to merge this and Prayer Flick into a 'Prayer' plugin. Or 'Prayer Indicators' as @deathbeam likes to call it 😄 

The plan was to add a tick indicator similar to the Prayer Flick plugin, that flashes over the bar, or something of that sort, and also address #2088 with that too.

**Comment by SoyChai** [comment]
Oh I completely misread the OP - I thought this *was* a prayer-flicking display, since you linked to it on #2088. I guess that's your ultimate goal though. Renaming the plugin to Prayer for this seems pretty clean.

File: runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarPlugin.java [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+			return;
+		}
+
+		if (config.hideIfOutOfCombat() && localPlayer.getHealth() == -1)
+		{
+			showingPrayerBar = false;
+		}
+	}
+
+	private boolean CheckIfAnyPrayerIsActive()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/prayerbar/PrayerBarPlugin.java]: `camelCase`

**Comment by deathbeam** [comment]
Needs rebase

**Pull request closed without merging**